### PR TITLE
Add Uniswap class and unit tests.

### PR DESCRIPTION
**Pull request by twalth3r**
This PR adds a basic Uniswap class to the python solver code. I decided to start simple and only write something that is specific to Uniswap - the abstraction layer for other pools can be added when needed.

Unit tests (using @marcovc's amazing randomized test framework :tada:):
```
python -m pytest tests/src/components/Uniswap_test.py
```

### CODE REVIEW

**Review by marcovc — APPROVED** [review]
Looks great. Love the test-fu!

File: python/tests/src/components/Uniswap_test.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    assert uni.balance1.balance >= 1 and uni.balance1.balance <= 10
+    assert uni.balance2.balance >= 1 and uni.balance2.balance <= 10
+
+
+@given(random_mixed_token_balance_pair())
+def test_constructor(tbp: Tuple[TB, TB]):
+    """Test Uniswap constructor."""
+    b1, b2 = tbp
+
+    delayed_ctor = partial(
```
Comment (josojo) [inline python/tests/src/components/Uniswap_test.py:31]: ctor is constructor, right?

If someones read the code for the first time, they would probably pause here and think about it. If you wanna make reading the code a little bit easier, you could spell it out.

File: python/tests/src/components/Uniswap_test.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            with pytest.raises(AssertionError):
+                uni1 == uni2
+        else:
+            assert uni1 == uni2
+    else:
+        assert uni1 != uni2
+
+
+@given(random_uniswap(), random_uniswap())
+def test_hashing(uni1: Uniswap, uni2: Uniswap):
```
Comment (josojo) [inline python/tests/src/components/Uniswap_test.py:69]: Is the testing so smart that it really verifies the path :
```
    if uni1 == uni2:
        assert len(s) == 1
```
as this is quite unlikely to happen if it was really on random.

**Review by josojo — APPROVED** [review]
Clean PR with good unit-testing 👍 

Exciting that we are starting this avenue as well, together with the cpp one.

The random testing is cool, though I am not sure whether it is appropriated in all tests.

File: python/tests/src/components/Uniswap_test.py (hunk 1 above)
Comment (twalth3r) [reply python/tests/src/components/Uniswap_test.py]: Changed in db8f5fe !

File: python/tests/src/components/Uniswap_test.py (hunk 2 above)
Comment (twalth3r) [reply python/tests/src/components/Uniswap_test.py:69]: I remember Marco mentioning that the test framework explicitly captures edge cases. @marcovc - can you confirm or am I wrong?

Comment (marcovc) [reply python/tests/src/components/Uniswap_test.py:69]: Not in the sense that it does some kind of introspection to the test, however it is smart in general, and will surely try a parameter combination where uni1==uni2.
